### PR TITLE
Add E2E Woo Subscriptions setup

### LINKED ISSUE #2521: Automate Woo Subscriptions setup on E2E test sites
To reduce the amount of manual work needed to use the E2E tests, we need to automate the installation of Woo Subscriptions in the test site as we do for the other requirements (test products, WooCommerce, the Stripe plugin release, etc.).

### PR DESCRIPTION
**Pull request by rafaelzaleski**
Closes #2521 

## Changes proposed in this Pull Request:

This PR streamlines the E2E testing process by automating the setup of WC Subscriptions in the environment. This eliminates the need for manual installation, ensuring smoother and more efficient testing.

## Testing instructions

1. Setup a new E2E environment on a JN site. Following the instructions [here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/tests/e2e/README.md).
2. There's a new variable in the `local.env` env file, `GITHUB_TOKEN`. This token should have access to the WC Subscriptions repository. The plugin will be downloaded from the latest release.
3. After running `npm run test:e2e-setup -- --base_url=SOME_URL_HERE`, manually check that the subscriptions plugin was installed on the test site and that it is working.
4. Run all the E2E tests for subscriptions (`npm run test:e2e -- --base_url=SOME_URL_HERE --grep @subscriptions`) and ensure they're working and that you didn't have to do any setup on WC Subscriptions.

---

-   [x] Covered with tests (or have a good reason not to test in description ☝️)
-   [x] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [x] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)

### CODE REVIEW

**Comment by ricardo** [comment]
Nice work! 👏 

I didn't fully review the code yet, but I tested the `test:e2e-setup` and `test:e2e`, and noticed a few things:

- When running `test:e2e-setup`, the script name made me think it would **only** do the setup, but instead, it does the setup **and** runs the tests.
- When running `test:e2e` with `--grep @subscriptions`, it will first reinstall the subscription plugin (one more time unexpectedly), and then run the subscriptions tests.

@rafaelzaleski Should we try to better separate the setup step from the actual tests?

- There seems to be an issue with the `woocommerce-blocks/subscription-product.spec`. It will fail for me for the `#phone` selector. It's actually `#billing-phone` for me. Not sure if it's a bug or something changed in the Blocks checkout.
- The setup was overall smooth, the only thing I struggled with was that I copied the URL of the JN site which comes with a trailing slash (https://secret-cow.jurassic.ninja/), and pasting that to `--base_url=` and `SSH_HOST`, that made the setup fail and took me some time to figure it out. This might be a common mistake for other people, so maybe we should try to ignore the trailing slash.

**Comment by rafaelzaleski** [comment]
> When running test:e2e-setup, the script name made me think it would only do the setup, but instead, it does the setup and runs the tests.
> ...
> Should we try to better separate the setup step from the actual tests?

@ricardo, I also don't like that the tests start after the setup, but I think this is work for another PR so I created [this issue](https://github.com/woocommerce/woocommerce-gateway-stripe/issues/2576).

I believe addressing the other points in this PR would be a good idea. I'll take a closer look and request another review once those points are covered. Thank you for pointing them out!

**Comment by rafaelzaleski** [comment]
@ricardo, can you try again?

> When running test:e2e with --grep @subscriptions, it will first reinstall the subscription plugin (one more time unexpectedly), and then run the subscriptions tests.

Now it only installs WC Subscriptions if `GITHUB_TOKEN` and `--with_woo_setup` are set. This only happens on `npm run test:e2e-setup`.

> There seems to be an issue with the woocommerce-blocks/subscription-product.spec. It will fail for me for the #phone selector. It's actually #billing-phone for me. Not sure if it's a bug or something changed in the Blocks checkout.

Fixed https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2522/commits/8cefc92ec24770e800708870708cd565f30da47a

> The setup was overall smooth, the only thing I struggled with was that I copied the URL of the JN site which comes with a trailing slash (https://secret-cow.jurassic.ninja/), and pasting that to --base_url= and SSH_HOST, that made the setup fail and took me some time to figure it out. This might be a common mistake for other people, so maybe we should try to ignore the trailing slash.

Fixed on https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2522/commits/a43b84c606d3d01f66108186ee5879f30702ab21 and https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2522/commits/c544a855b8b03cf4054469952f9e7dafaa1133d6

File: tests/e2e/env/e2e.sh [hunk 1]
```diff
@@ -25,6 +25,14 @@ for arg in "$@"; do
 
 done
 
+# Remove the latest trailing slash from the URL
+base_url=$(echo "$base_url" | sed 's:/*$::')
+
+# Helper to determine if we're running against Jurassic Ninja
+if [[ $base_url == *".jurassic.ninja" ]]; then
+  test_env="$test_env IS_JURASSIC='1'"
```
Comment (ricardo) [inline tests/e2e/env/e2e.sh]: > `IS_JURASSIC`

What is this used for? Should we call it a flag instead of a helper?

🤏 Also, this line specifically is indented with 2 spaces, the whole file is indented with spaces while the other `.sh` files are using tabs.

File: tests/e2e/env/e2e.sh [hunk 2]
```diff
@@ -25,6 +25,14 @@ for arg in "$@"; do
 
 done
 
+# Remove the latest trailing slash from the URL
```
Comment (ricardo) [inline tests/e2e/env/e2e.sh]: 🤏 A trailing slash = the latest `/`.

File: tests/e2e/utils/plugin-utils.js [hunk 3]
```diff
@@ -1,98 +1,75 @@
...
-			failOnStatusCode: true,
-		} );
+	// Make a request to the GitHub API to get information about the release
+	const releaseInfo = await axios( options ).catch( ( error ) => {
+		if ( error.response ) {
+			console.log( error.response.data );
+			console.log( error.response.status );
+			console.log( error.response.headers );
+		}
+		process.exit( 0 );
```
Comment (ricardo) [inline tests/e2e/utils/plugin-utils.js]: Are these logs useful, or a leftover for debugging? I guess we could be a little more generic, like:

```JavaScript
const releaseInfo = await axios( options ).catch( ( error ) => {
	console.error( `GitHub API request failed: ${error.message}` );
	process.exit( 1 );
} );
```

I guess we should be exiting with `0` when the node script was run successfully, otherwise, we should use `1` or a greater number.

File: tests/e2e/utils/plugin-utils.js [hunk 4]
```diff
@@ -107,43 +84,3 @@ export const deleteZip = async ( zipFilePath ) => {
...
 	console.log( `\u2714 Successfully deleted!` );
 };
-
-/**
- * Get the download URL of the release zip from GitHub
- *
- * @param string version The version to be tested.
- * @return string Download URL for the release zip file.
- */
-export const getReleaseZipUrl = async ( version ) => {
```
Comment (ricardo) [inline tests/e2e/utils/plugin-utils.js:117]: This function is still being used in [`global-teardown.js`](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/6ff08d33b76ab21a23c88248eb259cbce0efbe99/tests/e2e/config/global-teardown.js#L4).

File: tests/e2e/utils/pw-setup.js
Comment (ricardo) [inline tests/e2e/utils/pw-setup.js:1]: This is the second time I misread the `pw` in this file name as if it were related to "password" 😅.

This is supposed to stand for Playwright, correct? Maybe it's just me, or we could rename it to `playwright-setup.js`, in case it's not common to call it PW.

File: tests/e2e/utils/pw-setup.js [hunk 5]
```diff
@@ -189,6 +190,76 @@ export const installPluginFromRepository = ( page ) =>
 		} )();
 	} );
 
+/**
+ * Helper function to update the plugin.
```
Comment (ricardo) [inline tests/e2e/utils/pw-setup.js]: Maybe rephrase this to something like:

> Helper function to download and install the WooCommerce Subscriptions plugin from its latest GitHub release.

**Review by ricardo — CHANGES_REQUESTED** [review]
Tests well after the latest fix on `woocommerce-blocks/subscription-product.spec`. ✅ 

Only left a few minor comments, but code and approach LGTM.

File: tests/e2e/env/e2e.sh (hunk 1 above)
Comment (rafaelzaleski) [reply tests/e2e/env/e2e.sh]: I'll use this to get the SSH credentials for the JN site from the `wp-admin`, perhaps in another PR that separates the setup script from the tests.

Comment (rafaelzaleski) [reply tests/e2e/env/e2e.sh]: https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2522/commits/d08282004ddcbc2cf70c68bc0e497c87615425fe

File: tests/e2e/env/e2e.sh (hunk 2 above)
Comment (rafaelzaleski) [reply tests/e2e/env/e2e.sh]: https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2522/commits/d08282004ddcbc2cf70c68bc0e497c87615425fe

File: tests/e2e/utils/pw-setup.js
Comment (rafaelzaleski) [reply tests/e2e/utils/pw-setup.js:1]: No, I also think `playwright-setup.js` is better. I've changed the name.

File: tests/e2e/utils/plugin-utils.js (hunk 3 above)
Comment (rafaelzaleski) [reply tests/e2e/utils/plugin-utils.js]: Good catch 👍  https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2522/commits/13903fe11df8bd77efac04adda496dadfc6dec9e

File: tests/e2e/utils/pw-setup.js (hunk 5 above)
Comment (rafaelzaleski) [reply tests/e2e/utils/pw-setup.js]: I took the time to fix all docstrings on this file. https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2522/commits/7f9d4ed5e7fba1fafb1425bf1feb77dc8864854b

File: tests/e2e/utils/plugin-utils.js (hunk 4 above)
Comment (rafaelzaleski) [reply tests/e2e/utils/plugin-utils.js:117]: I've refactored the clean up of downloaded files [here](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2522/commits/9963871c5226e070a1e4f306f39dbb8075bc1b79). Now it'll delete the entire `tmp` folder after the tests instead of individual files.

File: tests/e2e/utils/playwright-setup.js [hunk 6]
```diff
@@ -122,7 +136,11 @@ export const createApiTokens = ( page ) =>
 	} );
 
 /**
- * Helper function to update the plugin.
+ * Helper function to download the Stripe plugin from the repository and install it on the site.
+ * This is useful when we want to test a specific version of the plugin.
+ * If the plugin is already installed, it will be updated to the specified version.
+ * @param {Page} page Playwright page object.
```
Comment (ricardo) [inline tests/e2e/utils/playwright-setup.js:142]: Feel free to ignore: In some places we're using `{Page} page`, and in other places `{Object} page`.

**Review by ricardo — APPROVED** [review]
LGTM. Tests well 👏
